include/engine/routing_algorithms/routing_base.hpp

```diff
@@ -217,84 +218,33 @@ template <class DataFacadeT, class Derived> class BasicRoutingInterface
-                if ((facade->GetTarget(edge_id) == edge.second) && (weight < edge_weight) &&
-                    facade->GetEdgeData(edge_id).forward)
-                {
-                    smaller_edge_id = edge_id;
-                    edge_weight = weight;
-                }
-            }
-
-            // edge.first         edge.second
-            //     *<------------------*
-            //            edge_id
-            // if we don't find a forward edge, this edge must have been an downwards edge
-            // found by the reverse search.
-            if (SPECIAL_EDGEID == smaller_edge_id)
-            {
-                for (const auto edge_id : facade->GetAdjacentEdgeRange(edge.second))
-                {
-                    const EdgeWeight weight = facade->GetEdgeData(edge_id).distance;
-                    if ((facade->GetTarget(edge_id) == edge.first) && (weight < edge_weight) &&
-                        facade->GetEdgeData(edge_id).backward)
-                    {
-                        smaller_edge_id = edge_id;
-                        edge_weight = weight;
-                    }
-                }
-            }
-            BOOST_ASSERT_MSG(edge_weight != INVALID_EDGE_WEIGHT, "edge id invalid");
-
-            const EdgeData &ed = facade->GetEdgeData(smaller_edge_id);
-            if (ed.shortcut)
-            { // unpack
-                const NodeID middle_node_id = ed.id;
-                // again, we need to this in reversed order
-                recursion_stack.emplace(middle_node_id, edge.second);
-                recursion_stack.emplace(edge.first, middle_node_id);
-            }
-            else
-            {
-                BOOST_ASSERT_MSG(!ed.shortcut, "original edge flagged as shortcut");
-                unsigned name_index = facade->GetNameIndexFromEdgeID(ed.id);
-                const auto turn_instruction = facade->GetTurnInstructionForEdgeID(ed.id);
+        UnpackCHEdge(
+            facade,
+            packed_path_begin,
+            packed_path_end,
+            [this,
+             &unpacked_path,
+             &phantom_node_pair,
+             &start_traversed_in_reverse,
+             &target_traversed_in_reverse](std::pair<NodeID, NodeID> & /* edge */,
+                                           const EdgeData &data) {
+
+                BOOST_ASSERT_MSG(!data.shortcut, "original edge flagged as shortcut");
+                unsigned name_index = facade->GetNameIndexFromEdgeID(data.id);
+                const auto turn_instruction = facade->GetTurnInstructionForEdgeID(data.id);
                 const extractor::TravelMode travel_mode =
                     (unpacked_path.empty() && start_traversed_in_reverse)
                         ? phantom_node_pair.source_phantom.backward_travel_mode
-                        : facade->GetTravelModeForEdgeID(ed.id);
+                        : facade->GetTravelModeForEdgeID(data.id);
```

Comment:
Nitpicking here, but data is a very general name. Would prefer something more verbose.